src/runtime/hexagon_dma_device_interface.cpp

```diff
@@ -0,0 +1,441 @@
+/**
+*  This file contains the definitions of the hexagon dma device interface class
+* This file is an interface between Halide Runtime and the DMA driver
+* The functions in this file take care of the various functionalities of the
+* DMA driver through their respective halide device interface functions
+* The function in this file use the Hexagon DMA context structure for sharing
+* the DMA context across various APIs of the
+* DMA device interface
+*/
+
+#include "runtime_internal.h"
+#include "device_buffer_utils.h"
+#include "device_interface.h"
+#include "printer.h"
+#include "hexagon_dma_device_shim.h"
+#include "HalideRuntimeHexagonDma.h"
+#include "hexagon_dma_rt.h"
+#include "hexagon_dma_context.h"
+
+extern "C" {
+
+extern WEAK halide_device_interface_t hexagon_dma_device_interface;
+
+WEAK int halide_hexagon_dma_device_release(void *user_context) {
+    return 0;
+}
+
+WEAK int halide_hexagon_dma_device_malloc(void *user_context,  halide_buffer_t *buf) {
+    int nRet = E_OK;
+    t_eDmaFmt chroma_type,luma_type;
+    int frame_idx;
+    t_dma_context* dma_context;
+    uintptr_t region_tcm;
+    uintptr_t region_tcm_desc;
+    uintptr_t tcm_buf_vaddr, tcm_desc_vaddr;
+    qurt_size_t region_tcm_desc_size;
+    qurt_mem_pool_t pool_tcm;
+
+    /*Get the Frame Index*/
+    if(halide_hexagon_dmart_get_frame_index(user_context, (uintptr_t)buf->host, &frame_idx)) {
+        error(NULL) << "Function failed to get Frame Index  \n";
```

Comment:
Pass user_context instead of NULL here and elsewhere, same for debug(NULL).

Reply:
Looking at more of the code, it seems like this might be problematic, since this code somewhat hijacks the user_context from its existing meaning.

You might want to take a look at the way OpenCL or CUDA works (e.g. https://github.com/halide/Halide/blob/master/src/runtime/opencl.cpp#L142), which is that we expose a function that is supposed to map user_context to the further OpenCL/CUDA context data. This way, the user can control the mapping between the user_context (the definition of which Halide does not control) and the dma_context.